Failed to Serialize Array of Bytes

The following log is obtained from publishing camera image data (which is array of bytes) using Kumo to a web app. It seems that there is a slight difference between serializing using array of bytes and other array of primitive types.

```
[ERROR] [1619250123.969132843] [subscription_5]: Failed to handle Subscription message! Object of type bytes is not JSON serializable
[ERROR] [1619250124.061008819] [subscription_5]: Failed to handle Subscription message! Object of type bytes is not JSON serializable
[ERROR] [1619250124.152951260] [subscription_5]: Failed to handle Subscription message! Object of type bytes is not JSON serializable
```

--- a/README.md
+++ b/README.md
@@ -1,6 +1,7 @@
 # Kumo JSON
 
 [![latest version](https://img.shields.io/github/v/release/ichiro-its/kumo_json)](https://github.com/ichiro-its/kumo_json/releases/)
+[![commits since latest version](https://img.shields.io/github/commits-since/ichiro-its/kumo_json/latest)](https://github.com/ichiro-its/kumo_json/commits/master)
 [![milestone](https://img.shields.io/github/milestones/progress/ichiro-its/kumo_json/1?label=milestone)](https://github.com/ichiro-its/kumo_json/milestone/1)
 [![license](https://img.shields.io/github/license/ichiro-its/kumo_json)](./LICENSE)
 [![test status](https://img.shields.io/github/workflow/status/ichiro-its/kumo_json/Build%20and%20Test?label=test)](https://github.com/ichiro-its/kumo_json/actions)

--- a/kumo_json/kumo_json/data_types.py
+++ b/kumo_json/kumo_json/data_types.py
@@ -46,7 +46,6 @@ def filter_type(data_type: str, value: any) -> any:
         value = float(value)
     elif is_byte(data_type):
         value = value.encode('ISO-8859-1')
-        print(value)
 
     return value
 

--- a/kumo_json/kumo_json/data_types.py
+++ b/kumo_json/kumo_json/data_types.py
@@ -45,7 +45,8 @@ def filter_type(data_type: str, value: any) -> any:
     elif is_float(data_type):
         value = float(value)
     elif is_byte(data_type):
-        value = str.encode(value)
+        value = value.encode('ISO-8859-1')
+        print(value)
 
     return value
 

--- a/kumo_json/kumo_json/data_types.py
+++ b/kumo_json/kumo_json/data_types.py
@@ -31,7 +31,7 @@ def is_float(data_type: str) -> bool:
     return data_type in ['float', 'double']
 
 
-def is_bytes(data_type: str) -> bool:
+def is_byte(data_type: str) -> bool:
     return data_type == 'octet'
 
 
@@ -44,7 +44,18 @@ def filter_type(data_type: str, value: any) -> any:
         value = int(value)
     elif is_float(data_type):
         value = float(value)
-    elif is_bytes(data_type):
+    elif is_byte(data_type):
         value = str.encode(value)
 
     return value
+
+
+def get_sequence_item_type(data_type: str) -> str:
+    opening_tag_index = data_type.find('<') + 1
+
+    for index, char in enumerate(data_type):
+        if char in [',', '>']:
+            closing_tag_index = index
+            break
+
+    return data_type[opening_tag_index:closing_tag_index]

--- a/kumo_json/kumo_json/json_to_msg.py
+++ b/kumo_json/kumo_json/json_to_msg.py
@@ -34,16 +34,9 @@ def dict_to_msg(msg_dict: dict, msg: MsgType) -> MsgType:
         value = msg_dict.get(field)
 
         if dtypes.is_array(data_type):
-            opening_tag_index = data_type.find('<') + 1
-
-            for index, char in enumerate(data_type):
-                if char in [',', '>']:
-                    closing_tag_index = index
-                    break
-
-            sequence_data_type = data_type[opening_tag_index:closing_tag_index]
+            sequence_item_type = dtypes.get_sequence_item_type(data_type)
             for index, item in enumerate(value):
-                value[index] = dtypes.filter_type(sequence_data_type, item)
+                value[index] = dtypes.filter_type(sequence_item_type, item)
         else:
             value = dtypes.filter_type(data_type, value)
 

--- a/kumo_json/kumo_json/msg_to_json.py
+++ b/kumo_json/kumo_json/msg_to_json.py
@@ -40,10 +40,14 @@ def msg_to_dict(msg: MsgType) -> dict:
             value = int(value)
         elif dtypes.is_float(data_type):
             value = float(value)
-        elif dtypes.is_bytes(data_type):
+        elif dtypes.is_byte(data_type):
             value = value.decode("utf-8")
         elif dtypes.is_array(data_type):
-            value = list(value)
+            if dtypes.is_byte(dtypes.get_sequence_item_type(data_type)):
+                value = [element.decode("utf-8") for element in value]
+            else:
+                value = list(value)
+        print(value)
 
         msg_dict[field] = value
 

--- a/kumo_json/kumo_json/msg_to_json.py
+++ b/kumo_json/kumo_json/msg_to_json.py
@@ -41,13 +41,12 @@ def msg_to_dict(msg: MsgType) -> dict:
         elif dtypes.is_float(data_type):
             value = float(value)
         elif dtypes.is_byte(data_type):
-            value = value.decode("utf-8")
+            value = value.decode('ISO-8859-1')
         elif dtypes.is_array(data_type):
             if dtypes.is_byte(dtypes.get_sequence_item_type(data_type)):
-                value = [element.decode("utf-8") for element in value]
+                value = [element.decode('ISO-8859-1') for element in value]
             else:
                 value = list(value)
-        print(value)
 
         msg_dict[field] = value
 

--- a/kumo_json/test/test_serialization/test_arrays_msg.py
+++ b/kumo_json/test/test_serialization/test_arrays_msg.py
@@ -32,6 +32,7 @@ def test_array_msg():
     msg.float32_array = [-3.1415, 3.1415]
     msg.bool_array = [True, False, True]
     msg.string_array = ["hello", "world", "!"]
+    msg.byte_array = [b'\x01', b'\x10', b'\x11']
 
     parsed_msg = json_to_msg(msg_to_json(msg), Arrays())
 
@@ -40,6 +41,7 @@ def test_array_msg():
     assert parsed_msg.float32_array == msg.float32_array
     assert parsed_msg.bool_array == msg.bool_array
     assert parsed_msg.string_array == msg.string_array
+    assert parsed_msg.byte_array == msg.byte_array
 
 
 def test_array_msg_from_json():
@@ -48,7 +50,8 @@ def test_array_msg_from_json():
                     "uint8_array": [1, 2, 3],
                     "float32_array": [-3.1415, 3.1415],
                     "bool_array": [true, false, true],
-                    "string_array": ["hello", "world", "!"]
+                    "string_array": ["hello", "world", "!"],
+                    "byte_array": ["\u0001", "\u0010", "\u0011"]
                   }'''
 
     parsed_msg = json_to_msg(msg_json, Arrays())
@@ -58,3 +61,4 @@ def test_array_msg_from_json():
     assert parsed_msg.float32_array == array('f', [-3.1415, 3.1415])
     assert parsed_msg.bool_array == [True, False, True]
     assert parsed_msg.string_array == ["hello", "world", "!"]
+    assert parsed_msg.byte_array == [b'\x01', b'\x10', b'\x11']

--- a/kumo_json_interfaces/msg/Arrays.msg
+++ b/kumo_json_interfaces/msg/Arrays.msg
@@ -3,3 +3,4 @@ uint8[] uint8_array
 float32[] float32_array
 bool[] bool_array
 string[] string_array
+byte[] byte_array

